### PR TITLE
MOTECH-3080: Fixed observations group

### PR DESCRIPTION
**Pull request by mkruszynski**

### CODE REVIEW

File: openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java [hunk 1]
```diff
@@ -913,52 +913,44 @@ private MotechEvent createEventForMember(String cohortQueryUuid, CohortQueryRepo
...
+
+        Concept concept5 = new Concept();
+        concept5.setUuid("parentUuid");
 
         Observation parentObservation = new Observation();
         parentObservation.setObsDatetime(DATE);
-        parentObservation.setConcept(concept6);
-        parentObservation.setGroupMembers(Arrays.asList(child1, child2, child3));
+        parentObservation.setConcept(concept5);
+        parentObservation.setGroupMembers(Arrays.asList(child1, child2, child3, child4));
```
Comment (kmadej) [inline openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java]: why 4 children not 3 ?

**Review by kmadej — CHANGES_REQUESTED** [review]

File: openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java (hunk 1 above)
Comment (mkruszynski) [reply openmrs/src/test/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceTest.java]: @kmadej OpenMRS, for some reason, is ignoring nested groupMembers. When I'm creating only the observations which are containing some value (last UUID in group), they're created.

File: openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSEncounterServiceImpl.java [hunk 2]
```diff
@@ -186,4 +197,39 @@ private void validateEncounter(Encounter encounter) {
...
+                        groupMembers.add(nestedObservation);
+                    }
+                }
+            }
+            observationToCreate.setGroupMembers(groupMembers);
+            observationToCreate.setEncounter(encounter);
+
+            Observation createdObservation = observationService.createObservation(configName, observationToCreate);
+
+            if (createdObservation.getGroupMembers() == null) {
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSEncounterServiceImpl.java]: this is redundant, it is already handled in getter

File: openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSEncounterServiceImpl.java [hunk 3]
```diff
@@ -186,4 +197,39 @@ private void validateEncounter(Encounter encounter) {
...
+        List<Observation> createdObservations = new ArrayList<>();
+        Observation observationToCreate;
+
+        for (Observation observation : observationList) {
+            List<Observation> nestedObservations = new ArrayList<>();
+            List<Observation> groupMembers = new ArrayList<>();
+
+            observationToCreate = observation;
+
+            if (observation.getGroupMembers() != null) {
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSEncounterServiceImpl.java]: redundant

File: openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSEncounterServiceImpl.java [hunk 4]
```diff
@@ -52,9 +59,13 @@ public Encounter createEncounter(String configName, Encounter encounter) {
 
         Encounter createdEncounter;
         Config config = configService.getConfigByName(configName);
+        List<Observation> observationList = encounter.getObs();
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSEncounterServiceImpl.java:66]: please add a comment to this code

File: openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSEncounterServiceImpl.java [hunk 5]
```diff
@@ -186,4 +197,39 @@ private void validateEncounter(Encounter encounter) {
 
         return result;
     }
+
+    private List<Observation> createNestedObservations(String configName, Encounter encounter, List<Observation> observationList) {
+        List<Observation> createdObservations = new ArrayList<>();
+        Observation observationToCreate;
+
+        for (Observation observation : observationList) {
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/service/impl/OpenMRSEncounterServiceImpl.java:209]: observationList can be null, please handle this in getObs() getter in Encounter

**Review by kmadej — CHANGES_REQUESTED** [review]

**Comment by kkedys** [comment]
@mkruszynski, I checked. The subgroups are not visible in OpenMRS. I verify it in API documentation. 

I created the observations concerning the following schema:

1. Observation with ANC Concept, without obs value and with the following children (obs groupsMembers)
a) Observation with Visit Month Concept, obs value = 3
b) Observation with ANC Visit Concept, obs value = 2nd
c) Observation with Diagnosis Concept, without obs value, with the following children (obs groupsMembers)
          Observation with High Risk Concept, obs value = No

The problem that I reached is the last observation (Observation with High Risk Concept, obs value = No) is not visible on OpenMRS.

**Comment by mkruszynski** [comment]
@kkedys I checked this for this concepts (every concept had Datatype - Text) and all observations was created successfully. Can you check this one more time?

**Comment by kkedys** [comment]
@kmadej , I checked one more time and any observations are visible in the OpenMRS.

**Comment by kmadej** [comment]
OpenMRS ITs fail. Additionally please add person to observation in EncounterService not in ActionProxyService

**Review by kmadej — APPROVED** [review]

**Pull request merged**